### PR TITLE
Add basic P2P to squad chess

### PR DESCRIPTION
**Pull request by EzraWeller**
Still need to add standard js and testing for the new p2p package.

### CODE REVIEW

File: packages/app-spec-squad-chess/src/Matchmaker.js [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+import settings from './settings.json'
+import state from './state.js'
+import { checkWinner } from './Board.js'
+
+export const Matchmaker = {
+  oninit: () => {
+    state.p2p.id = crypto.randomBytes(16).toString('hex')
+    console.log(`Our matchmaking Id: ${state.p2p.id}`)
+
+    state.p2p.room = 'Room name'
```
Comment (jessebmiller) [inline packages/app-spec-squad-chess/src/Matchmaker.js]: Why is this initialized to 'Room name'?

File: packages/app-spec-squad-chess/src/Matchmaker.js [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+  p2p.whenServerReady(async () => {
+    p2p.joinRoom(state.p2p.room)
+    p2p.listenOffers(handleReceiveOffer)
+    p2p.listenConnectionStatus(handleConnectionStatus)
+    p2p.listenMessage(handleReceiveMessage)
+    state.p2p.rollCallInterval = setInterval(async () => {
+      await rollCall()
+      m.redraw()
+    }, 1000)
+    p2p.whenServerReady(() => {})
```
Comment (jessebmiller) [inline packages/app-spec-squad-chess/src/Matchmaker.js]: Might be nice to console log that the server is ready

File: packages/p2p/app/index.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+const host = conf('PEER_DISCOVERY_HOST', 'localhost')
+const port = conf('PEER_DISCOVERY_PORT', '8889')
+
+const server = new WSServer({ host, port })
+
+console.log(`Peer discovery server listening on ws://${host}:${port}`)
+
+const ROOMS = {}
```
Comment (jessebmiller) [inline packages/p2p/app/index.js]: I would probably down case this. I generally reserve all caps for global constants that feel like configuration.

File: packages/p2p/app/index.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    return
+  }
+  console.log(`Id ${id} joining room ${room}`)
+  ROOMS[room].push(id)
+
+  // Time out from the room after 1 minute
+  setTimeout(() => {
+    console.log(`ID ${id} timed out from room ${room}`)
+    leaveRoom(room, id)
+  }, 240000)
```
Comment (jessebmiller) [inline packages/p2p/app/index.js]: Let's make this configurable, for now we can make a constant at the top of the file `const ROOM_TIMEOUT=240000`

File: packages/p2p/app/index.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+})
+
+server.register('rollCall', ([room]) => {
+  if (ROOMS[room]) {
+    return ROOMS[room]
+  }
+  return []
+})
+
+server.register('addEvent', ([event]) => {
```
Comment (jessebmiller) [inline packages/p2p/app/index.js:54]: Why is this event coming in as a list?

File: packages/p2p/app/index.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  if (!server.eventList().includes(event)) {
+    server.event(`${event}`)
+    console.log('Added event', event)
+  }
+})
+
+server.register('triggerEvent', ([event, data, from]) => {
```
Comment (jessebmiller) [inline packages/p2p/app/index.js:61]: Why are these params coming in as a list?

File: packages/p2p/browser-test/src/index.js [hunk 7]
```diff
@@ -0,0 +1,224 @@
+import m from "mithril"
+import matchmaking from "../../clients/js"
+
+let connectionStatus = 'closed'
+let messages = []
+const room = 'squadChess'
```
Comment (jessebmiller) [inline packages/p2p/browser-test/src/index.js:6]: Let's all caps room here to `ROOM`. below when I saw the variable `room` I was confused about where it came from, but if it was formatted as a global config style variable that would have been clear.

File: packages/p2p/clients/js/index.js [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  ourId = userId
+
+  server = null
+  theirId = null
+  room = 'empty'
+  survivingPeer = null
+  offeringPeer = null
+  answeringPeer = null
+  dataChannel = null
+  connectionStatusCB = null
```
Comment (jessebmiller) [inline packages/p2p/clients/js/index.js]: Let's move these to a `resetState` function

File: packages/p2p/clients/js/index.js [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+  listenEvent('offer', callback)
+  listenEvent('answer', (e) => {
+    acceptAnswer(e.from, e.data)
+  })
+  listenEvent('candidate', (e) => {
+    addCandidate(e.from, e.data)
+  })
+}
+
+function listenEvent(eventType, callback) {
```
Comment (jessebmiller) [inline packages/p2p/clients/js/index.js]: Is there a reason we are renaming `subscribe` and `uncubscribe` to `listen` and `stopListen`. If we can we should try to use the established language

File: packages/p2p/clients/js/index.js [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+  events[eventType] = eventName(eventType, ourId)
+  server.call('addEvent', [events[eventType]])
+  server.subscribe(events[eventType])
+  server.on(events[eventType], async (e) => {
+    callback(e)
+  })
+}
+
+function stopListen(eventType) {
```
Comment (jessebmiller) [inline packages/p2p/clients/js/index.js]: Name this to match the form of its pair. 
so `listenEvent` and `stopListenEvent`
or `subscribeEvent` and `unsubscribeEvent`

File: packages/p2p/clients/js/index.js [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+async function sendAnswer(id, offer) {
+  // leave the matchmaking system & start the connecting process
+  offeringPeer = null
+  leaveRoom()
+  stopListen('offer')
+  stopListen('answer')
+  survivingPeer = answeringPeer
+  theirId = id
```
Comment (jessebmiller) [inline packages/p2p/clients/js/index.js:147]: Why not name it `theirId` in the function deffinition?

Comment (jessebmiller) [reply packages/p2p/clients/js/index.js:147]: oh never mind I see. This confusion does highlight that we might not want to use the global scope for all the state variables. Might be better to put them in a single object.

File: packages/p2p/clients/js/index.js [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+  if (id != theirId) { throw new Error(`Candidate ID ${id} does not match target id ${theirId}`)}
+  console.log(`Adding candidate ${candidate} from ID ${id}`)
+  survivingPeer.addIceCandidate(candidate).catch(e => { throw new Error(e) })
+}
+
+function sendMessage(msg) {
+  dataChannel.send(msg)
+}
+
+/*** CALLBACKS ***/
```
Comment (jessebmiller) [inline packages/p2p/clients/js/index.js]: I would categorize these after their function (handlers) rather than their use as callbacks.

**Review by jessebmiller — CHANGES_REQUESTED** [review]
This is a ton of great work! Just got a few notes.

Super excited to see p2p squad chess in action!

**Comment by jessebmiller** [comment]
Oh also, does this code style get checked in CI?

**Comment by EzraWeller** [comment]
Nope, not yet -- I ran it manually once, but I didn't add it to the CI yet. I can try to figure out how to do that as well.

File: packages/app-spec-squad-chess/src/Matchmaker.js (hunk 1 above)
Comment (EzraWeller) [reply packages/app-spec-squad-chess/src/Matchmaker.js]: It was getting prefilled from here into the form, but I don't think there's a reason to do it that way, changing it.

File: packages/p2p/app/index.js (hunk 5 above)
Comment (EzraWeller) [reply packages/p2p/app/index.js:54]: This is was the way it was presented in the documentation for rpc-websockets -- I'll check if it works without the list.

Comment (EzraWeller) [reply packages/p2p/app/index.js:54]: Yup: `TypeError: "params" must be an object or "Array"`

File: packages/p2p/clients/js/index.js (hunk 10 above)
Comment (EzraWeller) [reply packages/p2p/clients/js/index.js]: Just going to go with `subscribe` and `unsubscribe` I think.

File: packages/p2p/clients/js/index.js (hunk 11 above)
Comment (EzraWeller) [reply packages/p2p/clients/js/index.js:147]: I thought about that, and ended up deciding it wasn't worth the extra readability bloat of a state object, but it does seem like it's on the edge. Down to change it if you feel strongly.

**Pull request closed without merging**